### PR TITLE
Change loss function to a unique_ptr

### PR DESCRIPTION
**Pull request by jackgerrits**
- Removes several raw deletes in the codebase
- Additionally this change moves to scope based control in nn

### CODE REVIEW

File: vowpalwabbit/nn.cc [hunk 1]
```diff
@@ -262,7 +260,7 @@ void predict_or_learn_multi(nn& n, single_learner& base, example& ec)
       }
     }
 
-    n.all->loss = save_loss;
+    loss_function_swap_guard_converse_block.do_swap();
```
Comment (lalo) [inline vowpalwabbit/nn.cc:263]: do we need all the do_swaps technically? since they will go out of scope and trigger eventually, this might be more explicit though

**Comment by lalo** [comment]
might be worthwhile to add a test that uses swap guard and unique_ptr in case impl of swap_guard changes? not sure how probable though https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/test/unit_test/guard_test.cc

File: vowpalwabbit/nn.cc (hunk 1 above)
Comment (jackgerrits) [reply vowpalwabbit/nn.cc:263]: Perhaps not, but this code is pretty complicated and makes use of a `goto` so I didn't want to mess with the logic. This should be equivalent to the original code.

**Comment by jackgerrits** [comment]
Yep good thinking, can do!

**Review by lalo — APPROVED** [review]

**Review by olgavrou — APPROVED** [review]
